src/Microsoft.ML.Data/StaticPipe/Reconciler.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.ML.Core.Data;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+
+namespace Microsoft.ML.Data.StaticPipe.Runtime
+{
+    /// <summary>
+    /// An object for <see cref="PipelineColumn"/> instances to indicate to the analysis code for static pipelines that
+    /// they should be considered a single group of columns (through equality on the reconcilers), as well as how to
+    /// actually create the underlying dynamic structures, whether an <see cref="IDataReaderEstimator{TSource, TReader}"/>
+    /// (for the <see cref="ReaderReconciler{TREaderIn}"/>) or a <see cref="IEstimator{TTransformer}"/>
+    /// (for the <see cref="EstimatorReconciler"/>).
+    /// </summary>
+    public abstract class Reconciler
+    {
+        private protected Reconciler() { }
+    }
+
+    /// <summary>
+    /// Reconciler for column groups intended to resolve to a new <see cref="IDataReaderEstimator{TSource, TReader}"/>
+    /// or <see cref="IDataReader{TSource}"/>.
+    /// </summary>
+    /// <typeparam name="TIn">The input type of the <see cref="IDataReaderEstimator{TSource, TReader}"/>
+    /// object.</typeparam>
+    public abstract class ReaderReconciler<TIn> : Reconciler
+    {
+        public ReaderReconciler() : base() { }
```

Comment:
(nit - both here and for EstimatorReconciler) - typically `public` constructors on `abstract` classes are frowned upon.  These should be `protected`. See [https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1012-abstract-types-should-not-have-constructors](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1012-abstract-types-should-not-have-constructors)